src/Oceananigans.jl

```diff
@@ -132,7 +132,48 @@ end
 
 abstract type Architecture end
 struct CPU <: Architecture end
-struct GPU <: Architecture end
+
+# A slightly complex (but fully-featured?) implementation:
+
+struct Layout{NT, NB}
+    threads :: NTuple{NT, Int}
+     blocks :: NTuple{NB, Int}
+end
+
+XYZLayout(threads, grid) = Layout(threads, 
+    (floor(Int, threads[1]/grid.Nx), floor(Int, threads[2]/grid.Ny), grid.Nz) )

```

Comment:
We'll have to do something more sophisticated here as you there is a maximum number of threads you can have. We found that it's roughly a bit more than ~1024 for a V100 so thread-block layouts must have that `Tx*Ty*Tz <= 1024`.

I think right now we use `Tx, Ty = 16, 16` so we're effectively saying 16^2 = 256 is the maximum number of threads (but if you schedule enough blocks you'll use up all the threads on the GPU).

Reply:
I just wrote in a default of `(16, 16)`.

Perhaps you can advise how one calculates the block sizes associated with a particular thread layout in a general way?

Reply:
> Perhaps you can advise how one calculates the block sizes associated with a particular thread layout in a general way?

So I think the idea is to choose the number of threads `Tx, Ty, Tz` up to some maximum number of threads, maybe `max_threads = 1024`. Then you create `Bx*By*Bz` blocks, each containing `Tx*Ty*Tz` threads such that each thread ends up calculating stuff for 1 grid point.

But we also have to make sure that we can create `Bx*By*Bz` blocks so the number of threads has to evenly divide the number of grid points. This is easy for nice domains, e.g. `256^3` but can probably be challenging for weird domain sizes, e.g. `112x223x71`.

Reply:
sounds like a problem in factorization. I'm not sure its actually challenging. There simply will be domains for which a layout cannot be calculated, and we will throw an error in that case.

Reply:
I see a problem. The kernels seem to make an assumption about a relationship between the axes of the data arrays and the size of the blocks. It seems that generalizing this problem will involve rewriting *all* of the kernels... ?

Reply:
Yeah I guess it's a rather trivial problem. I'm sure you could still come up with a valid thread block layout for `112x223x71`.

I suppose it's a question of whether we want to support weird domain sizes. For now, we probably shouldn't worry too much about it.

Reply:
Why do we not ever use `Tz` right now? Our domains are typically cubic.

Reply:
I guess when I initially started running on GPUs I noticed that we would saturate the number of threads per block using just `Tx, Ty` even on a moderate 32^3 grid so I didn't go up to `Tz`.

Not sure if this was a mistake or not, but the benchmarking I've done convinced me that you don't need `Tz` to achieve good performance.